### ISSUE TITLE
GLM warmup_dflash.py samples at temperature 0 / thinking off: first temperature>0 request after readiness JITs _gumbel_sample_kernel (and fused_recurrent_kda_fwd_kernel on a 4K-prompt burst)

### ISSUE THREAD

**Issue by sethforprivacy** [FujitsuPolycom/sparkring#214]
## Summary

`runtime/glm53-flash-jj-r8-gb10/warmup_dflash.py` sends every readiness request with `"temperature": 0.0` and `"enable_thinking": False`. The documented serving profile is temperature 1.0 with thinking on, so the first real request after every launch still triggers a Triton JIT for the sampling path, after Docker has reported the rank healthy and the router has admitted traffic.

## Observed (image `0d4029b3…` / id `5e32aaa1…`, launcher at `efe404d`, four GB10 ranks, TP4/DCP4, `DFLASH_WARMUP=1`, `DFLASH_WARMUP_CONCURRENCIES=1..16`)

Rank-0 `jit_monitor` (mode `warn`), same lines on all four ranks:

```
21:45:53 Kernel JIT monitor activated; monitored JIT compilations during inference will use mode=warn.
21:46:30-31  (during the warm-up itself) _expand_c4_block_table_kernel, _pool_seq_lens_kernel, _prepare_dflash_inputs_kernel,
             _compute_local_logits_stats_kernel, _rejection_kernel, _resample_kernel, _prefill_pool_kernel,
             _prefill_tail_kernel, _gather_c4_block_table_rows_kernel
21:50:27 Triton kernel JIT compilation during inference: fused_recurrent_kda_fwd_kernel   <- 8 concurrent ~4K-token prompts behind a running 133K request (our SparkCache skip-path gate), after readiness
21:59:07 Triton kernel JIT compilation during inference: _gumbel_sample_kernel             <- first temperature 0.7 request of the day on this lane, after readiness
```

On this launch the JIT namespace was warm (second launch on `glm53-flash-sm121-vllm-e02b1746-b12x-9ae41c5c`), and the Triton cache directories for both kernels predate the warnings (`_gumbel_sample_kernel` written 21:33Z, `fused_recurrent_kda_fwd_kernel` 21:45Z), so these two were disk-cache loads and cost little. On the cold first launch of this composition (13:08Z today, new source-bound namespace) the same two kernels were full compiles during live traffic; the 13:08Z as-run only caught `fused_recurrent_kda_fwd_kernel` because the gates run at temperature 0 as well. `performance/records/glm53-flash/dflash2-exact-concurrency-graphs-20260904.md` already notes the KDA kernel compiling "while the C5 8K stream was becoming ready".

## Ask

1. Add one warm-up request at `temperature: 1.0` (any concurrency; the profile's documented sampling mode) so `_gumbel_sample_kernel` is compiled before readiness.
2. Consider one warm-up request with `enable_thinking: true` so the reasoning-parser path is exercised before readiness (no kernel involved, but it is the profile default and currently never runs before health).
3. Add a mixed-shape burst (several ~4K-token prompts while a longer request decodes) to cover `fused_recurrent_kda_fwd_kernel`, or document the shape that specialises it.
4. Optional: have `jit_monitor` distinguish a Triton disk-cache load from a compile (or log the elapsed time), so operators can tell a 100 ms load from a multi-second compile on a cold namespace.

Happy to test a branch on the four-rank site.

**Comment by sethforprivacy** [comment]
Update after running the published `20260906-indexer-barrier-warmup` image (`@sha256:28e6a9c0…` / id `d2ba5894…`) on the same four-GB10 TP4/DCP4 lane: **ask 1 is implemented, and it does not close the gap.**

`serve-with-warmup.py` in that image does send the readiness request at `"temperature": 1.0` with `"chat_template_kwargs": {"enable_thinking": True}`, before rank 0 reports healthy. Confirmed in the running container. (`warmup_dflash.py` is unchanged at `temperature: 0.0` / `enable_thinking: False`, which is fine if the new request is meant to be the sampling-path cover.)

But `_gumbel_sample_kernel` still compiles after readiness, on the first request that sets its own sampling parameters.

**Run A — cold JIT namespace** (`glm53-20260906-indexer-barrier-warmup`, fresh per the hotfix README), 2026-09-06: readiness completed, rank 0 healthy, gates green. The first request afterwards with an explicit `temperature: 0.7` triggered
`Triton kernel JIT compilation during inference: _gumbel_sample_kernel` at 18:58:42Z. It was **the only post-readiness JIT event in the entire run** — every other kernel was covered by the warmup.

**Run B — same image, second launch on the now-warm namespace**, corroborating which kernels the readiness request actually exercises. The monitor (mode `warn`, active 19:15:28Z, i.e. before the warmup) reports this set at 19:16:09–19:16:31Z:

```
_expand_c4_block_table_kernel, _pool_seq_lens_kernel, _prepare_dflash_inputs_kernel,
_compute_local_logits_stats_kernel, _rejection_kernel, _resample_kernel,
_prefill_pool_kernel, _prefill_tail_kernel, _gather_c4_block_table_rows_kernel,
fused_recurrent_kda_fwd_kernel
```

`_gumbel_sample_kernel` is **not** in it. Three subsequent requests at the server's default sampling settings also did not trigger it — so the kernel is reached by a branch the temperature-1.0 readiness request does not take, rather than by "any temperature > 0".

I have not isolated the discriminating axis and do not want to guess in your tracker; the two candidates from our side are the sampler branch selected when a client supplies explicit sampling parameters (our ladder sends `temperature: 0.7`, no `top_p`/`top_k`) versus a shape specialization (streaming, concurrency, `max_tokens: 300`). Both are cheap for you to check against the sampler than for us to infer from the outside.

**Suggested amendment to ask 1:** warm the shape real clients send, not just a non-zero temperature — one additional readiness request with explicit `temperature` plus `top_p`/`top_k`, streamed. On a cold namespace the current warmup leaves exactly one kernel compiling under live traffic after the router has been told the rank is ready.

Minor note for anyone reading the timestamps: this JIT monitor warns on disk-cache loads as well as true compiles, so a warning on a warm namespace is cheap. The cost that matters is the cold-namespace case, which the fresh-namespace instruction in the hotfix README makes the normal path for anyone adopting that image.

**Comment by sethforprivacy** [comment]
Follow-up on the unisolated axis, plus a status correction from our side.

**Prior art: a second GB10 kit treats this as a sampler-tensor-combination problem.** The MiaAI EXL3 kit ships [`scripts/boot-shape-warmup.sh`](https://github.com/MiaAI-Lab/GLM-5.3-Flash-EXL3-2x-DGX-Sparks/blob/main/scripts/boot-shape-warmup.sh) (`GLM53_BOOT_SHAPE_WARMUP`), which burns DFlash2 / sampler / kpool shapes after `/health` and sweeps three explicit sampler arms rather than one non-zero temperature:

```
sampling-k)  '"top_k":40,"top_p":1.0'
sampling-p)  '"top_k":0,"top_p":0.9'
*)           '"top_k":40,"top_p":0.9'
```

Their stated rationale, which is the part relevant here:

> Hub `generation_config.json` stamps `top_p=0.95`. Omitting `top_p` on a k-only arm compiles TOPK+TOPP, never k-only. `top_p=1.0` / `top_k=0` are how glm53-flash's sampler drops the p / k tensors (None).

If that holds generally, the discriminator is **which sampler tensors survive to the kernel as non-None**, not "temperature > 0" — which is candidate 1 from my previous comment, narrowed. They also assert coverage rather than inferring it: `verify_sampler_cache()` inspects the Triton cache for the expected `_topk_topp_kernel` constexpr combos and fails the postcondition if any are missing. That design seems worth borrowing regardless of the root cause, since it turns "did the warmup cover the sampler?" into a check instead of an assumption.

**Caveat that matters for sparkring specifically: the default source is checkpoint-dependent.** Their mechanism depends on the checkpoint stamping `top_p`. The checkpoint sparkring serves on our lane does not. On `local-inference-lab/GLM-5.3-Flash-NVFP4-Spark` @ `df116c4f…`, `generation_config.json` is `_from_model_config: true` with **every sampling field null** — `temperature`, `top_k`, `top_p`, `min_p`, `typical_p` all `null`; the only non-null entries are `use_cache`, `pad_token_id`, `eos_token_id`, `output_attentions`, `output_hidden_states`. So on this checkpoint a client sending bare `temperature: 0.7` is not picking up a stamped `top_p`, and their specific explanation does not transfer as-is. A warmup hardcoded to one arm will not generalize across the checkpoints sparkring serves; sweeping the combinations would.

**Status correction: I can no longer reproduce this, and it is not because it was fixed.** The lane moved off the DFlash2 profile to the native-MTP3 mesh composition (`sparkring-glm53-sparkcache`, cache+checkpoints, upstream #236) on 2026-09-07. On that image the post-readiness gap does not appear: every `jit_monitor` warning falls inside a ~35 s window during the warmup sweep that follows `Application startup complete` —

```
_expand_c4_block_table_kernel, _pool_seq_lens_kernel, _compute_local_logits_stats_kernel,
_rejection_kernel, _resample_kernel, _prefill_pool_kernel, _prefill_tail_kernel,
_gather_c4_block_table_rows_kernel, _pack_topk_routes_small_prefix_kernel,
_pack_topk_routes_sort_kernel, _store_cache_checkpoints_kernel
```

— last one at 21:06:07Z, followed by **427 chat completions with zero JIT events**. `_gumbel_sample_kernel` does not appear anywhere in the run.

I am not reading that as the issue being resolved. The warmup sweep on this image is still entirely `temperature: 1.0` (visible in its own `dflash_warmup` output), so the k-only / p-only arms are unexercised here rather than proven covered — no client on this lane has sent them. It does mean I cannot run the discriminating A/B for you any more, which is why I am pointing at someone else's implementation instead of finishing the isolation myself.

**Suggested amendment to ask 1, unchanged in spirit but concrete:** if the readiness warmup is going to cover the sampling path, sweep the sampler tensor combinations (k-only, p-only, both) rather than a single non-zero temperature, and gate readiness on a cache postcondition rather than on the requests having been sent. `boot-shape-warmup.sh` is a working reference for both halves.

Credit to Mia's AI Lab for the isolation; that repo is AGPL-3.0, so it is a reference here, not a code drop.

**Comment by FujitsuPolycom** [comment]
Readiness warmup now covers the sampling-path arms requested in the amendment: `runtime/glm53-flash-jj-r8-gb10/serve_with_warmup.py` `SAMPLING_CASES` sends six streamed requests (unfiltered, temperature 0.7, top-k only, top-p only, top-k+top-p, seeded top-k+top-p) with `enable_thinking: True` before rank 0 reports healthy.

Left open of the amendment: (1) a mixed-shape burst covering `fused_recurrent_kda_fwd_kernel` specialization (the single non-zero-temperature request is not the full tensor-combination sweep for the recurrent kernel); (2) `jit_monitor` distinguishing Triton disk-cache loads from full compiles; (3) the cache-postcondition gate asserting the expected constexpr combinations are present rather than readiness on the assumption that the request was sent. The discriminator identified in the thread — which sampler tensors reach the kernel as non-None — is unconfirmed and stays the open question behind the coverage claims.
